File: src/agent/open-canvas/nodes/customAction.ts
import { ChatOpenAI } from "@langchain/openai";
import { OpenCanvasGraphAnnotation, OpenCanvasGraphReturnType } from "../state";
import {
  CUSTOM_QUICK_ACTION_ARTIFACT_CONTENT_PROMPT,
  CUSTOM_QUICK_ACTION_ARTIFACT_PROMPT_PREFIX,
  CUSTOM_QUICK_ACTION_CONVERSATION_CONTEXT,
  REFLECTIONS_QUICK_ACTION_PROMPT,
} from "../prompts";
import { ensureStoreInConfig, formatReflections } from "../../utils";
import {
  ArtifactContent,
  CustomQuickAction,
  Reflections,
} from "../../../types";
import { LangGraphRunnableConfig } from "@langchain/langgraph";
import { BaseMessage } from "@langchain/core/messages";

const formatMessages = (messages: BaseMessage[]): string =>
  messages
    .map(
      (msg) =>
        `<${msg.getType()}>\n${msg.content as string}\n</${msg.getType()}>`
    )
    .join("\n");

export const customAction = async (
  state: typeof OpenCanvasGraphAnnotation.State,
  config: LangGraphRunnableConfig
): Promise<OpenCanvasGraphReturnType> => {
  if (!state.customQuickActionId) {
    throw new Error("No custom quick action ID found.");
  }

  const smallModel = new ChatOpenAI({
    model: "gpt-4o-mini",
    temperature: 0.5,
  });

  const store = ensureStoreInConfig(config);
  const assistantId = config.configurable?.assistant_id;
  if (!assistantId) {
    throw new Error("`assistant_id` not found in configurable");
  }
  const customActionsNamespace = ["custom_actions", assistantId];
  const actionsKey = "actions";

  const memoryNamespace = ["memories", assistantId];
  const memoryKey = "reflection";

  const [customActionsItem, memories] = await Promise.all([
    store.get(customActionsNamespace, actionsKey),
    store.get(memoryNamespace, memoryKey),
  ]);
  if (!customActionsItem?.value) {
    throw new Error("No custom actions found.");
  }
  const customQuickAction = customActionsItem.value[
    state.customQuickActionId
  ] as CustomQuickAction | undefined;
  if (!customQuickAction) {
    throw new Error(
      `No custom quick action found from ID ${state.customQuickActionId}`
    );
  }

  let currentArtifactContent: ArtifactContent | undefined;
  if (state.artifact) {
    currentArtifactContent = state.artifact.contents.find(
      (art) => art.index === state.artifact.currentContentIndex
    );
  }
  if (!currentArtifactContent) {
    throw new Error("No artifact content found.");
  }

  let formattedPrompt = `<custom-instructions>\n${customQuickAction.prompt}\n</custom-instructions>`;
  if (customQuickAction.includeReflections && memories?.value) {
    const memoriesAsString = formatReflections(memories.value as Reflections);
    const reflectionsPrompt = REFLECTIONS_QUICK_ACTION_PROMPT.replace(
      "{reflections}",
      memoriesAsString
    );
    formattedPrompt += `\n\n${reflectionsPrompt}`;
  }

  if (customQuickAction.includePrefix) {
    formattedPrompt = `${CUSTOM_QUICK_ACTION_ARTIFACT_PROMPT_PREFIX}\n\n${formattedPrompt}`;
  }

  if (customQuickAction.includeRecentHistory) {
    const formattedConversationHistory =
      CUSTOM_QUICK_ACTION_CONVERSATION_CONTEXT.replace(
        "{conversation}",
        formatMessages(state.messages.slice(-5))
      );
    formattedPrompt += `\n\n${formattedConversationHistory}`;
  }

  formattedPrompt += `\n\n${CUSTOM_QUICK_ACTION_ARTIFACT_CONTENT_PROMPT.replace("{artifactContent}", currentArtifactContent.content)}`;

  const newArtifactValues = await smallModel.invoke([
    { role: "user", content: formattedPrompt },
  ]);

  const newArtifact = {
    ...state.artifact,
    currentContentIndex: state.artifact.contents.length + 1,
    contents: [
      ...state.artifact.contents,
      {
        ...currentArtifactContent,
        index: state.artifact.contents.length + 1,
        content: newArtifactValues.content as string,
      },
    ],
  };

  return {
    artifact: newArtifact,
  };
};

File: agents-py/agents/src/open_canvas/nodes/custom_action.py
from typing import Optional, Dict, Any
from uuid import uuid4
from langchain_core.messages import BaseMessage
from langchain_core.runnables import RunnableConfig
from agents.src.open_canvas.state import OpenCanvasGraphState
from shared.src.types import (
    ArtifactCodeV3,
    ArtifactMarkdownV3,
    ArtifactV3,
    CustomQuickAction,
    Reflections
)
from agents.src.utils import (
    ensure_store_in_config,
    format_reflections,
    get_model_from_config
)
from shared.src.prompts.quick_actions import (
    CUSTOM_QUICK_ACTION_ARTIFACT_CONTENT_PROMPT,
    CUSTOM_QUICK_ACTION_ARTIFACT_PROMPT_PREFIX,
    CUSTOM_QUICK_ACTION_CONVERSATION_CONTEXT,
    REFLECTIONS_QUICK_ACTION_PROMPT
)

def format_messages(messages: list[BaseMessage]) -> str:
    return "\n".join(
        f"<{msg.type}>\n{msg.content}\n</{msg.type}>"
        for msg in messages
    )

async def custom_action(
    state: OpenCanvasGraphState,
    config: RunnableConfig
) -> Dict[str, Any]:
    if not state.custom_quick_action_id:
        raise ValueError("No custom quick action ID found")

    small_model = await get_model_from_config(config, {"temperature": 0.5})
    
    store = ensure_store_in_config(config)
    assistant_id = config.get("configurable", {}).get("assistant_id")
    user_id = config.get("configurable", {}).get("supabase_user_id")
    
    if not assistant_id:
        raise ValueError("`assistant_id` not found in configurable")
    if not user_id:
        raise ValueError("`user.id` not found in configurable")

    # Parallel fetching of data
    custom_actions_namespace = ["custom_actions", user_id]
    memory_namespace = ["memories", assistant_id]
    
    custom_actions_item, memories = await (
        store.get(custom_actions_namespace, "actions"),
        store.get(memory_namespace, "reflection")
    )

    if not custom_actions_item or "value" not in custom_actions_item:
        raise ValueError("No custom actions found")
    
    action_id = state.custom_quick_action_id
    custom_action = custom_actions_item["value"].get(action_id)
    if not custom_action:
        raise ValueError(f"No custom quick action found for ID {action_id}")

    current_artifact_content = None
    if state.artifact:
        current_artifact_content = state.artifact.contents[-1]  # Simplified from getArtifactContent

    # Build formatted prompt
    formatted_prompt = f"<custom-instructions>\n{custom_action.prompt}\n</custom-instructions>"
    
    if custom_action.include_reflections and memories and "value" in memories:
        reflections_str = format_reflections(memories["value"])
        formatted_prompt += f"\n\n{REFLECTIONS_QUICK_ACTION_PROMPT.format(reflections=reflections_str)}"
    
    if custom_action.include_prefix:
        formatted_prompt = f"{CUSTOM_QUICK_ACTION_ARTIFACT_PROMPT_PREFIX}\n\n{formatted_prompt}"
    
    if custom_action.include_recent_history:
        conv_history = CUSTOM_QUICK_ACTION_CONVERSATION_CONTEXT.format(
            conversation=format_messages(state._messages[-5:])
        )
        formatted_prompt += f"\n\n{conv_history}"

    artifact_content = getattr(current_artifact_content, "code", None) if current_artifact_content else None
    formatted_prompt += f"\n\n{CUSTOM_QUICK_ACTION_ARTIFACT_CONTENT_PROMPT.format(artifact_content=artifact_content or 'No artifacts generated yet.')}"

    # Invoke model
    response = await small_model.invoke([{"role": "user", "content": formatted_prompt}])

    if not current_artifact_content:
        print("No current artifact content found")
        return {}

    # Create new artifact content
    new_content = response.content
    new_artifact_content = {
        **current_artifact_content.dict(),
        "index": len(state.artifact.contents) + 1,
        "code": new_content if isinstance(current_artifact_content, ArtifactCodeV3) else None,
        "full_markdown": new_content if isinstance(current_artifact_content, ArtifactMarkdownV3) else None
    }

    # Create new artifact
    new_artifact = ArtifactV3(
        current_index=len(state.artifact.contents) + 1,
        contents=[*state.artifact.contents, new_artifact_content]
    )

    return {"artifact": new_artifact} 